src/ber/mod.rs

```diff
@@ -0,0 +1,367 @@
+#[derive(Clone, Copy, Debug)]
+#[repr(u8)]
+pub enum Class {
+    Universal = 0,
+    Application = 1,
+    ContextSpecific = 2,
+    Private = 3,
+}
+
+/// Represents whether or not the type is Primitive or Constructed
+#[derive(Clone, Copy, Debug)]
+#[repr(u8)]
+pub enum Aspect {
+    Primitive = 0,
+    Constructed = 1,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Tag(u8);
+
+impl Tag {
+    /// Construct a new `Tag`
+    pub const fn new(byte: u8) -> Self {
+        Tag(byte)
+    }
+
+    pub const fn from_parts(class: Class, aspect: Aspect, tag: u8) -> Self {
+        Tag((tag & 0x1F) | ((class as u8) << 6) | ((aspect as u8) << 5))
+    }
+
+    /// Get the `Class` of the tag
+    pub fn class(self) -> Class {
+        match self.0 >> 6 {
+            0 => Class::Universal,
+            1 => Class::Application,
+            2 => Class::ContextSpecific,
+            3 => Class::Private,
+            _ => unreachable!("matching on two bits"),
+        }
+    }
+
+    /// Get the `Aspect` of the tag
+    pub fn aspect(self) -> Aspect {
+        match (self.0 >> 5) & 1 {
+            0 => Aspect::Primitive,
+            1 => Aspect::Constructed,
+            _ => unreachable!("matching on one bit"),
+        }
+    }
+
+    /// Get the tag number
+    pub fn number(self) -> u8 {
+        self.0 & 0x1F
+    }
+
+    /// Get the raw value of the tag
+    pub const fn raw(self) -> u8 {
+        self.0
+    }
+}
+
+impl Serialize for Tag {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        buffer.write_byte(self.raw());
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct Length(u64);
+
+impl Length {
+    pub fn new(len: u64) -> Self {
+        Self(len)
+    }
+}
+
+impl Deserialize for Length {
+    fn deserialize(bytes: &mut &[u8]) -> Result<Self, DeserializeError> {
+        const MASK_LENGTH_DEFINITE: u8 = 0b1000_0000;
+        const LENGTH_DEFINITE_SHORT: u8 = 0b0000_0000;
+
+        let length_type = bytes.byte()?;
+        let length_value = u64::from(length_type & 0x7F);
+
+        match length_type & MASK_LENGTH_DEFINITE {
+            LENGTH_DEFINITE_SHORT => Ok(Self::new(length_value)),
+            _ if length_value > 0 => Ok(Self::new(bytes.uint(Length::new(length_value))?)),
+            _ => Err(DeserializeError::IndefiniteLength),
+        }
+    }
+}
+
+impl Serialize for Length {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
+        if self.0 > 127 {
+            let ns = self.0.to_be_bytes();
+            let start = ns.iter().copied().position(|n| n != 0).unwrap_or(7);
+            buffer.write_byte(0x80 | (8 - start as u8));
+            buffer.write(&ns[start..]);
+        } else {
+            buffer.write_byte(self.0 as u8);
+        }
+    }
+}
+
+impl std::ops::Add for Length {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl Serialize for str {
+    fn serialize(&self, buffer: &mut dyn VecExt) {
```

Comment:
i think this could work with generics instead of dyn dispatch

Reply:
I don't think it really matters to be honest

Reply:
it doesn't , but it could make things simpler. Honestly just a tidbit